mindmeld/components/dialogue.py

```diff
@@ -689,6 +692,249 @@ def _get_dialogue_handler(self, dialogue_state):
         return handler
 
 
+class AutoEntityFilling:
+    """A special dialogue flow sublcass to implement Automatic Entitiy (Slot) Filling
+    (AEF) that allows developers to prompt users for completing the missing
+    requirements for entity slots.
+
+    Attributes:
+        app (Application): The application that initializes this flow.
+    """
+
+    _logger = mod_logger.getChild("AutoEntityFilling")
+    """Class logger."""
+
+    def __init__(self, entrance_handler, form, app):
+        self._app = app
+        self._entrance_handler = entrance_handler
+        self._form = form
+        self._local_form = None
+        self._prompt_user = None
+        self._check_attr()
+
+    def _check_attr(self):
+        if not ('entities' in self._form and len(self._form['entities']) > 0):
+            raise KeyError("Entity list cannot be empty.")
+
+        self._entity_form = self._form['entities']
+        self._max_retries = (
+            self._form['max_retries'] if 'max_retries' in self._form else 1)
+        self._exit_response = (
+            self._form['exit_msg'] if 'exit_msg' in self._form else 'How may I help you?')
+        self._exit_intent = (
+            self._form['exit_intent'] if 'exit_intent' in self._form else None)
+        self._exit_keys = (
+            self._form['exit_keys'] if 'exit_keys' in self._form else
+            ['cancel', 'restart', 'exit', 'reset'])
+
+        if not isinstance(self._max_retries, int):
+                    raise TypeError("'max_retries' should be of type: int.")
+        if not isinstance(self._exit_response, str):
+                    raise TypeError("'exit_msg' should be of type: str.")
+        if self._exit_intent and not isinstance(self._exit_intent, list):
+                    raise TypeError("'exit_intent' should be of type: list.")
+        if not isinstance(self._exit_keys, list):
+            raise TypeError("'exit_keys' should be of type: list.")
+
+        self._exit_keys = list(map(str.lower, self._exit_keys))
+
+    def _set_target_state(self, request, responder):
+        """Set target dialogue state to the entrance handler's name"""
+        responder.params.allowed_intents = tuple([''.join([request.domain, ".", request.intent])])
+        responder.params.target_dialogue_state = self._entrance_handler.__name__
+
+    def _exit_flow(self, responder, intents=None):
+        """Exits this flow and clears the related parameter for re-usability"""
+        self._prompt_user = None
+        self._local_form = None
+        responder.params.allowed_intents = tuple(intents) if intents else tuple()
```

Comment:
This is setting the next allowed_intents to be the exit intents, regardless of what the user is saying, which is not what we want. We should remove the exit_intents for now.

```suggestion
        responder.params.allowed_intents = tuple(intents) if intents else tuple()
```